internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -0,0 +1,179 @@
+package rabbitpubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/google/go-cloud/internal/pubsub"
+	"github.com/google/go-cloud/internal/pubsub/driver"
+	"github.com/streadway/amqp"
+)
+
+type topic struct {
+	exchange string // the AMQP exchange
+	conn     *amqp.Connection
+
+	mu   sync.Mutex
+	ch   *amqp.Channel            // AMQP channel used for all communication.
+	pubc <-chan amqp.Confirmation // Go channel for server acks of publishes
+}
+
+// Values for the amqp client.
+// See https://www.rabbitmq.com/amqp-0-9-1-reference.html.
+const (
+	// Many methods of the amqp client take a "no-wait" parameter, which
+	// if true causes the client to return without waiting for a server
+	// response. We always want to wait.
+	wait = false
+
+	// Always use the empty routing key. This driver expects to be used with topic
+	// exchanges, which disregard the routing key.
+	routingKey = ""
+
+	// If the message can't be enqueued, return it to the sender rather than silently dropping it.
+	mandatory = true
+
+	// If there are no waiting consumers, enqueue the message instead of dropping it
+	immediate = true
+)
+
+// OpenTopic returns a *pubsub.Topic corresponding to the named exchange. The
+// exchange must have been previously created (for instance, by using
+// amqp.Channel.ExchangeDeclare). For the model of Go Cloud Pub/Sub to make sense,
+// the exchange should be a fanout exchange, although nothing in this package
+// enforces that.
+//
+// OpenTopic uses the supplied amqp.Connection for all communication. It is the caller's
+// responsibility to establish this connection before calling OpenTopic, and to close
+// it when it is no longer in use.
+//
+// The documentation of the amqp package recommends using separate connections for
+// publishing and subscribing.
+func OpenTopic(conn *amqp.Connection, name string) (*pubsub.Topic, error) {
+	// TODO(jba): support context.Context
+	t, err := newTopic(conn, name)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(t), nil
+
+}
+
+func newTopic(conn *amqp.Connection, name string) (*topic, error) {
+	t := &topic{
+		exchange: name,
+		conn:     conn,
+	}
+	if err := t.establishChannel(); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// establishChannel creates an AMQP channel if necessary. According to the amqp
+// package docs, once an error is returned from the channel, it must be discarded and
+// a new one created.
+//
+// Must be called with the t.mu held.
```

Comment:
Nit: The call to this function in `newTopic` does not hold `mu`. Probably fine, but contradicts this comment.

Reply:
Fixed by grabbing mu. (Probably unnecessary, but simpler.)